cvmfs/swissknife_list_reflog.cc

```diff
@@ -0,0 +1,177 @@
+    return 1;
+  }
+
+  const bool follow_redirects = false;
+  if (!this->InitDownloadManager(follow_redirects) ||
+      !this->InitVerifyingSignatureManager(repo_keys)) {
+    LogCvmfs(kLogCatalog, kLogStderr, "failed to init repo connection");
+    return 1;
+  }
+
+  bool success;
+  if (IsHttpUrl(repo_url)) {
+    HttpObjectFetcher<> object_fetcher(repo_name,
+                               repo_url,
+                               temp_directory,
+                               download_manager(),
+                               signature_manager());
+    success = Run(&object_fetcher, repo_name, output_path, reflog_hash);
+  } else {
+    LocalObjectFetcher<> object_fetcher(repo_url, temp_directory);
+    success = Run(&object_fetcher, repo_name, output_path, reflog_hash);
+  }
+
+  return (success) ? 0 : 1;
+}
+
+template <class ObjectFetcherT>
+bool CommandListReflog::Run(ObjectFetcherT *object_fetcher, string repo_name,
+                            string output_path, shash::Any reflog_hash)
+{
+  typename ObjectFetcherT::ReflogTN *reflog;
+  reflog = FetchReflog(object_fetcher, repo_name, reflog_hash);
+
+  // Traverse through catalogs and regular objects
+  vector<shash::Any> catalogs;
+  if (NULL == reflog || !reflog->List(SqlReflog::kRefCatalog, &catalogs)) {
+    LogCvmfs(kLogCvmfs, kLogStderr, "Failed to list catalog reference log");
+    return false;
+  }
+  typename CatalogTraversal<ObjectFetcherT>::Parameters traversal_params;
+  traversal_params.object_fetcher = object_fetcher;
+  CatalogTraversal<ObjectFetcherT> traversal(traversal_params);
+  traversal.RegisterListener(
+    &swissknife::CommandListReflog::CatalogCallback,
+    this);
+  bool success = true;
+  vector<shash::Any>::iterator i = catalogs.begin();
+  const vector<shash::Any>::const_iterator iend = catalogs.end();
+  for (; i != iend && success; i++) {
+    success &= traversal.TraverseRevision(*i,
+                CatalogTraversal<ObjectFetcherT>::kBreadthFirstTraversal);
+  }
+
+  if (!success) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+             "Catalog traversal aborted due to an error");
+  }
+
+  // Add history, certificate, metainfo objects from reflog
+  vector<shash::Any> histories, certificates, metainfos;
+  if (!reflog->List(SqlReflog::kRefHistory, &histories)) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+             "Failed to fetch history objects from reflog");
+  }
+  if (!reflog->List(SqlReflog::kRefCertificate, &certificates)) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+             "Failed to fetch certificate objects from reflog");
+  }
+  if (!reflog->List(SqlReflog::kRefMetainfo, &metainfos)) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+             "Failed to fetch metainfo objects from reflog");
+  }
+  objects_.insert(histories.begin(), histories.end());
+  objects_.insert(certificates.begin(), certificates.end());
+  objects_.insert(metainfos.begin(), metainfos.end());
+
+  // Clean up reflog file
+  reflog->DropDatabaseFileOwnership();
```

Comment:
Can't we let the reflog keep the database ownership, so that it takes care internally of the file cleanup?